webapp/src/debugger.tsx

```diff
@@ -0,0 +1,243 @@
+import * as React from "react";
+import * as ReactDOM from 'react-dom';
+import * as sui from "./sui";
+import * as data from "./data";
+import * as simulator from "./simulator";
+
+type ISettingsProps = pxt.editor.ISettingsProps;
+
+interface DebuggerVariablesState {
+    variables?: pxt.Map<Variable>;
+}
+
+interface Variable {
+    value: string;
+    type: string;
+    prevValue?: string;
+}
+
+interface DebuggerVariablesProps extends ISettingsProps {
+}
+
+export class DebuggerVariables extends data.Component<DebuggerVariablesProps, DebuggerVariablesState> {
+
+    private nextVariables: pxt.Map<pxsim.Variables> = {};
+
+    constructor(props: DebuggerVariablesProps) {
+        super(props);
+        this.state = {
+            variables: {}
+        }
+    }
+
+    clear() {
+        //this.setState({ variables: {} });
+    }
+
+    set(name: string, value: pxsim.Variables) {
+        this.nextVariables[name] = value;
+    }
+
+    update() {
+        const variables = this.state.variables;
+        Object.keys(this.nextVariables).forEach(k => {
+            const v = this.nextVariables[k];
+            let sv = '';
+            let type = typeof (v);
+            switch (type) {
+                case "number": sv = v + ""; break;
+                case "boolean": sv = v + ""; break;
+                case "string": sv = JSON.stringify(v); break;
+                case "object":
+                    if (v == null) sv = "null";
+                    else if (v.id !== undefined) sv = "(object)"
+                    else if (v.text) sv = v.text;
+                    else sv = "(unknown)"
+                    break;
+            }
+            variables[k] = {
+                value: sv,
+                type: type,
+                prevValue: variables[k] && sv != variables[k].value ?
+                    variables[k].value : undefined
+            }
+        })
+        this.setState({ variables: variables });
+        this.nextVariables = {};
+    }
+
+    renderCore() {
+        const { variables } = this.state;
+        return Object.keys(variables).length == 0 ? <div /> :
+            <div className="ui segment debugvariables">
+                <div className="ui middle aligned list">
+                    {Object.keys(variables).map(variable =>
+                        <div key={variable} className="item">
+                            <div className="ui label image variable" style={{ backgroundColor: pxt.blocks.getNamespaceColor('variables') }}>
+                                <span className="varname">{variable}</span>
+                                <div className="detail">
+                                    <span className="varval">{variables[variable].value + ' '}</span>
+                                    <span className="previousval">{variables[variable].prevValue ? '(' + variables[variable].prevValue + ')' : ''}</span>
```

Comment:
IMO remove variables.

Reply:
What!? I thought that's the primary feature you've been wanting.. I don't understand.

Reply:
Sorry wrong comment, how about remove parenthesis?

Reply:
Haha you had me worried for a sec.